src/KilnUniV3.sol

```diff
@@ -0,0 +1,123 @@
+// SPDX-FileCopyrightText: © 2022 Dai Foundation <www.daifoundation.org>
+// SPDX-License-Identifier: AGPL-3.0-or-later
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.8.14;
+
+import {KilnBase, GemLike} from "./KilnBase.sol";
+import {TwapProduct}       from "./uniV3/TwapProduct.sol";
+
+// https://github.com/Uniswap/v3-periphery/blob/b06959dd01f5999aa93e1dc530fe573c7bb295f6/contracts/SwapRouter.sol
+interface SwapRouterLike {
+    function exactInput(ExactInputParams calldata params) external returns (uint256 amountOut);
```

Comment:
`exactInput` is defined as `external' in the contract but in [UniSwap's interfaces](https://github.com/Uniswap/v3-periphery/blob/b06959dd01f5999aa93e1dc530fe573c7bb295f6/contracts/interfaces/ISwapRouter.sol#L37) it is `external payable`. Not sure if this will be an issue but worth matching if this is the corresponding interface.

```suggestion
    function exactInput(ExactInputParams calldata params) external payable returns (uint256 amountOut);
```

Reply:
The `payable` keyword is not part of the formal function signature, and in this case not having it in the interface is better for safety as there is no intention or support for transferring raw eth (so trying to do that will fail already in compilation). So I think we can leave this as is.

Reply:
No problem, this is a mistake on my part.